dataset/test/bins_view_test.cpp

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+#include <gtest/gtest.h>
+
+#include "test_macros.h"
+
+#include "scipp/dataset/bins.h"
+#include "scipp/dataset/bins_view.h"
+#include "scipp/dataset/dataset.h"
+#include "scipp/dataset/except.h"
+
+using namespace scipp;
+using namespace scipp::dataset;
+
+class BinsViewTest : public ::testing::Test {
+protected:
+  Dimensions dims{Dim::Y, 2};
+  Variable indices = makeVariable<scipp::index_pair>(
+      dims, Values{std::pair{0, 2}, std::pair{2, 4}});
+  Variable data =
+      makeVariable<double>(Dims{Dim::X}, Shape{4}, Values{1, 2, 3, 4});
+  DataArray buffer = DataArray(data, {{Dim::X, data + data}});
+  Variable var = make_bins(indices, Dim::X, buffer);
+};
+
+TEST_F(BinsViewTest, erase) {
+  auto view = bins_view<DataArray>(var);
+  auto da = var.bin_buffer<DataArray>();
+  EXPECT_TRUE(da.coords().contains(Dim::X));
+  view.coords().erase(Dim::X);
+  EXPECT_FALSE(da.coords().contains(Dim::X));
+}
+
+TEST_F(BinsViewTest, insert) {
+  auto view = bins_view<DataArray>(var);
+  const auto &da = var.bin_buffer<DataArray>();
+  EXPECT_FALSE(da.coords().contains(Dim::Y));
+  view.coords().set(Dim::Y, view.coords()[Dim::X]);
+  EXPECT_TRUE(da.coords().contains(Dim::Y));
+}
+
+TEST_F(BinsViewTest, slice_readonly) {
+  auto slice = var.slice({Dim::Y, 0});
+  auto view = bins_view<DataArray>(slice);
+  ASSERT_THROW(view.coords().erase(Dim::X), except::DataArrayError);
+  auto buffer = slice.bin_buffer<DataArray>();
+  ASSERT_THROW(buffer.coords().erase(Dim::X), except::DataArrayError);
+  EXPECT_TRUE(buffer.is_readonly());
+  EXPECT_TRUE(buffer.coords().is_readonly());
+  EXPECT_TRUE(buffer.masks().is_readonly());
+  EXPECT_TRUE(buffer.attrs().is_readonly());
+  EXPECT_TRUE(buffer.meta().is_readonly());
```

Comment:
Presumably there is no need to additionally check that a subsequent copy of the slice and view made from that would give mutable behaviour.

Reply:
At that point it is just a `DataArray` like any other, i.e., as you say (and unrelated to this change) the readonly flags (of the dicts but not of the data) would be dropped. I'll nevertheless add a test for clarity.